akka-stream/src/main/scala/akka/stream/impl/UnfoldResourceSourceAsync.scala

```diff
@@ -4,138 +4,140 @@
-    create: () => Future[S],
-    readData: (S) => Future[Option[T]],
-    close: (S) => Future[Done])
+@InternalApi private[akka] final class UnfoldResourceSourceAsync[R, T](
+    create: () => Future[R],
+    readData: (R) => Future[Option[T]],
+    close: (R) => Future[Done])
     extends GraphStage[SourceShape[T]] {
   val out = Outlet[T]("UnfoldResourceSourceAsync.out")
   override val shape = SourceShape(out)
   override def initialAttributes: Attributes = DefaultAttributes.unfoldResourceSourceAsync
 
-  def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with OutHandler {
-    lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
-    private implicit def ec: ExecutionContext = materializer.executionContext
-    private var state: Option[S] = None
-
-    private val createdCallback = getAsyncCallback[Try[S]] {
-      case Success(resource) =>
-        state = Some(resource)
-        if (isAvailable(out)) onPull()
-      case Failure(t) => failStage(t)
-    }.invokeWithFeedback _
-
-    private val errorHandler: PartialFunction[Throwable, Unit] = {
-      case NonFatal(ex) =>
-        decider(ex) match {
-          case Supervision.Stop =>
-            failStage(ex)
-          case Supervision.Restart =>
-            try {
-              restartResource()
-            } catch {
-              case NonFatal(ex) => failStage(ex)
-            }
-          case Supervision.Resume => onPull()
-        }
-    }
-
-    private val readCallback = getAsyncCallback[Try[Option[T]]](handle).invoke _
-
-    private def handle(result: Try[Option[T]]): Unit = result match {
-      case Success(data) =>
-        data match {
-          case Some(d) => push(out, d)
-          case None    =>
-            // end of resource reached, lets close it
-            state match {
-              case Some(resource) =>
-                close(resource).onComplete(getAsyncCallback[Try[Done]] {
-                  case Success(Done) => completeStage()
-                  case Failure(ex)   => failStage(ex)
-                }.invoke)(parasitic)
-                state = None
-
-              case None =>
-                // cannot happen, but for good measure
-                throw new IllegalStateException("Reached end of data but there is no open resource")
-            }
-        }
-      case Failure(t) => errorHandler(t)
-    }
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic with OutHandler =
+    new GraphStageLogic(shape) with OutHandler {
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      private implicit def ec: ExecutionContext = materializer.executionContext
+      private var maybeResource: OptionVal[R] = OptionVal.none
+
+      private val createdCallback = getAsyncCallback[Try[R]] {
+        case Success(resource) =>
+          maybeResource = OptionVal(resource)
```

Comment:
This will be a small change in behavior. it will not work when resource is null, which was probably ok previously.
Should we have a null check with a clear error message and say that null is not allowed?

Reply:
Oops I see!